### DISCUSSION TITLE
Contributing

### DISCUSSION THREAD

**Discussion by y2k04** [IoeCmcomc/MCDatapacker#15] [General]
How do I contribute new data pack versions to this project? And how would I compile this?

**Comment by IoeCmcomc** [comment]
To compile this project, you need a C++17 toolchain and Qt 1.15.2. The following tutorial assumes you're on Windows.
1. Download Qt [here](https://www.qt.io/download-qt-installer-oss).
2. Open the Qt Maintainance Tool. You need a Qt account to log in to this tool. Then, install the following components:
  - Qt 5.15.2:
    - MinGW 8.1.0 64-bit (if you haven't installed any C++ toolchain yet)
	- MinGW 8.1.0 32-bit (if your computer is 32-bit)
	- MSVC 2019 64-bit (if you have installed MSVC or Visual Studio)
	- MSVC 2019 32-bit (same as above, for 32-bit computers)
	- Sources
  - Developer and Designer Tools:
    - Qt Creator 12.0.2 (or newer)
	- Qt Creator 12.0.2 CDB Debugger Support
	- MinGW 8.1.0 64-bit (this is a C++ toolchain)
	- MinGW 8.1.0 32-bit (if your computer is 32-bit)
	- CMake 3.27.7 (or newer)
	- Ninja 1.10.2 (or newer)
	- OpenSSL 1.1.1v Toolkit
3. Fork this repository
4. **Clone the `dev` branch** using the following command:
   - `git clone https://github.com/<your_username>/MCDatapacker.git --recurse-submodules -b dev --single-branch`
   - The the time of writing, you will likely get errors if you try to clone the `master` branch.
4. Run Qt Creator to open `MCDatapacker.pro` (or double-click the file to open) the project. 
5. Choose a kit or all kits in the Configure Project screen.
6. Use buttons in the bottom left corner to build and run the project. When you run a project, Qt Creator will build the project before running it.

You also need Python 3.8+ to run scripts in the `tools` folder.

> How do I contribute new data pack versions to this project?

Well, it's complicated. I'm afraid you'll need to figure it out yourself because I'm rewriting some parts of the program, and there isn't an easy way to define custom Minecraft versions. However, you should look at the following files and folders:
- `resource` folder:
	- `minecraft` folder:
		- `info` folder:
			- 1.x.x: each folder in the form of 1.x.x (e.g. 1.20) contains various information about a particular Minecraft version.
				- `1.x.x.qrc` defines which files are embedded into (and available to) the program.
				- `registries` and `summary` folders are submodules pointing to 1.x.x registries from [Misode's `mcmeta`](https://github.com/misode/mcmeta "Misode's `mcmeta`") repository. 
				- Other JSON files are called "game info" files, which are either created manually or generated with the scripts in the `tools` folder.
		- `texture`: these game texture/icon files are shared between game versions.
- `src` folder:
	- `game.h`: this file defines constants for all Minecraft versions and defines functions to load game information. *A valid version must have the form "x.x.x"* (e.g. 1.20 or 1.20.4).
	- `newdatapackdialog.cpp`: mappings from Minecraft versions to pack format numbers are in this file.
	d

To initialize `registries` and `summary` folders:
1. Open a command line at the `resource/minecraft/info/1.x.x` folder.
2. Run `git submodule add --depth 1 https://github.com/misode/mcmeta.git summary`
3. Run `cd summary`
4. Run `git fetch origin tag 1.x.x-summary --no-tags -f`
5. Run `git checkout 1.20.4-summary`
6. Repeat steps 1 to 5 for the `registries` folder.

If a Minecraft version introduces new [command argument types](https://minecraft.wiki/w/Argument_types), you may need to add the new types to these files:
- `src` folder:
	- `codepalettle.h`: defines enum values of various syntax-highlighting elements in source codes.
	- `codepalettle.cpp`: contains two palettes (light and dark) and defines colors and formattings of highlighting elements.
	- `parsers/command` folder:
		- `nodes` folder: define AST-like parse node to represent the code of a Minecraft command or a mcfunction file.
			- `argumentnode.h`: contains enum values for argument types.
		- `schema` folder:
			- `schemaargumentnode.cpp`: define a mapping from an argument type's ID to its enum value defined in `argumentnode.h`.
		- `visitors` folder:
			- `nodevisitor.h`: the base visitor class.
			- `overloadnodevisitor.h`: implement default behaviors for not-defined visit methods.
			`reprprinter.h`: convert a parse tree to a debug string, used for debugging and testing purposes.
			`nodeformatter.h`: convert a parse tree to a list of format ranges that are used for syntax highlighting.
			`sourceprinter.h`: convert a parse tree to the code form. It's the opposite of parsing.
			`completionprovider.h`: provide code suggestions from an argument node.

If a Minecraft version introduces new "categories" (e.g. `recipes`, `damage_type`), you may need to change these files:
- `src` folder:
	- `codefile.h`: define all kinds of files which MCDatapacker can recognize.
	- `globalhelpers.cpp`: contains functions dealing with these "categories".
	- `datapacktreeview.cpp`: define how the context menu of the left-side tree view is created.

Last but not least, if a Minecraft version changes the structure of existing "categories", you need to update the corresponding side editors as well.

**Reply by y2k04** [reply]
Thank you!
